Add more solvation search tools with more advanced options

yunsiechung opened the pull request:
### **Overview **

The following solvation search tools have been added: 

1. **Solvent Search**: to search Abraham and Mintz solvent parameters and other solvent-specific parameters (e.g. viscosity) from RMG-database.
2. **Solute Parameter Search & Prediction (optional solvation free energy, enthalpy, and entropy prediction using the LSER)**: to search for Abraham solute parameters (E, S, A, B, L, V) from RMG-database or get the prediction from a group additivity model (SoluteGC) and a machine learning model (SoluteML) by entering a solute SMILES(s). Optionally, get solvation free energy, enthalpy, and entropy prediction at 298 K from the solute and solvent parameters using the LSER (Linear Solvation Energy Relationship) approach.
3. **Solvation Free Energy, Enthalpy, and Entropy Prediction at 298 K Using a Machine Learning Model**: to get solvation free energy, enthalpy, and entropy prediction at 298 K from a machine learning model (DirectML) by entering SMILES of a solvent and a solute. LogK and logP (dry) prediction is also available.
4. **Temperature-Dependent Solvation Free Energy, Vapor–Liquid Equilibria, and Henry's Law Constant Prediction**: to get solvation free energy, VLE, and Henry's law constant prediction along the solvent's saturation curve at any temperature of interest by entering SMILES of a solvent and a solute. LogK and logP (dry) prediction is also available. This temperature-dependence search is currently limited to certain solvents.

### **Important Notes **

Two new conda packages need to be installed for this PR, which is added to `environment.yml`. Please update the rmg environment by running `conda env update -f environment.yml` command.
You also have to make sure that you are no the most recent RMG-Py and RMG-database master branches.

### **Details **

**1. Solvent Search**

Any species identifier can be used to search for a solvent from RMG-database solvent library. Only one solvent can be input at a time. The results will display all solvent data that matched the input species. If the input species matches any of the solvents of the mixture solvent entry, the mixture solvent entry will also be matched. (e.g. water will match the acetonitrile & water mixture as well). If the input species is not available in the solvent library, it will display the message that it's not available.

Test this tool by entering the following examples:
- example 1: `water` -> will return 4 matched entries (`water`, `methanol_50_water_50`, `acetonitrile_40_water_60`, `acetonitrile_60_water_40`)
- example 2: `1,2-dichloroethene` -> will return 2 matched entries (`cis-1,2-dichloroethene`, `trans-1,2-dichloroethene`)
- example 3: `carbazole` -> will display the message `The given solvent is not found in the RMG solvent library.`.


**2. Solute Parameter Search & Prediction**

Details on this tool is explained on the page. (`/database/solvation/soluteSearch/`). A user can search or get predictions of solute parameters and their corresponding solvation energy, enthalpy and entropy at 298 K estimated from the LSER by entering a solute SMILES. The user can input up to 100 solute SMILES at once. Each SMILES needs to separated by empty space or line. The user can select a solute parameter estimation method among (1) RMG-database, (2) SoluteGC, and (3) SoluteML options.
A user can select a solvent if they want the optional solvation energy, enthalpy, and entropy search. A user can also select the preferred energy unit for the results.

If the user input more than 100 SMILES, it will display an error message `The number of input solute SMILESs cannot exceed 100`.
If the user does not select any estimation method for the solute parameter, it will display an error message `You must select an estimation method for the solute parameters`.
If the input format is wrong or SMILES cannot be parsed correctly, the result pages will display appropriate error messages. 

The results are displayed in a pandas html table format, and this result can be exported as an excel file by clicking `Click to Export to Excel` button in the result page.

Test this by entering the following input:
  `[H][H]  C(CBr)O c1ccccc1 C(C)(=O)O C1=CC=CC2=CC=CC=C12 lllllll Mg`
Choose any solute parameter estimation method and any solvent you want. 
If you choose 'SoluteML' as a estimation method and 'water' as a solvent, you will get a table showing results below:

![image](https://user-images.githubusercontent.com/17207530/126550086-ceeeda7b-286d-42df-9aed-5d0f14617339.png)
![image](https://user-images.githubusercontent.com/17207530/126549822-7aadc535-5639-45d6-9bd8-9cc4fa66c336.png)



**3. Solvation Free Energy, Enthalpy, and Entropy Prediction at 298 K Using a Machine Learning Model**

Details on this tool is explained on the page. (`/database/solvation/searchML/`). A user can get 298 K predictions of (1) solvation free energy (2) solvation enthalpy (3) solvation entropy (4) logK, and (5) logP by entering SMILES of a solvent and solute. A user can select any of those 5 options they want for calculations. A user can also select the preferred energy unit for the results.

The user can input up to 200 solvent-solute pairs for each search. The SMILES of a solvent and solute compound needs to be separated by the underscore (_) and each solvent-solute pair is separated by line or empty space. 

If the user input more than 200 solvent-solute pairs, it will display an error message `The number of input solvent-solute pairs cannot exceed 200`.
If no option is selected for calculation, it will display an error message `At least one option must be selected`.
If the input format is wrong or SMILES cannot be compiled correctly, the result pages will display appropriate error messages. 

The results are displayed in a pandas html table format, and this result can be exported as an excel file by clicking `Click to Export to Excel` button in the result page.

Test this by entering the following input:
```
CC1=NC=CC=C1_CCCCCCCC
CC(CC(C)C)=O_Oc1ccccc1
c1ccccc1_Cn1cccc1
C[S](C)=O_CCCC
lala
yuno_CCC
```
and choose any options you want. You will get a table showing results below:

![image](https://user-images.githubusercontent.com/17207530/126549955-6cbfc347-8417-4a5e-afd3-6d516d164e4f.png)



**4. Temperature-Dependent Solvation Free Energy, Vapor–Liquid Equilibria, and Henry's Law Constant Prediction**

Details on this tool is explained on the page. (`/database/solvation/searchTempDep/`). A user can get predictions of (1) solvation free energy (2) K-factor (3) Henry's constant (4) logK, and (5) logP by entering SMILES of a solvent and solute and a temperature of interest. A user can select any of those 5 options they want for calculations. A user can specify the unit of the input temperature. A user can also select the preferred energy unit for the results.

The user can input up to 200 solvent-solute-temperature input for each search. The SMILES of a solvent and solute compound and temperature needs to be separated by the underscore (_) and each solvent-solute-temperature input is separated by line or empty space. 

If the user input more than 200 solvent-solute-temperature input, it will display an error message `The number of solvent-solute-temperature inputs cannot exceed 200`.
If no option is selected for calculation, it will display an error message `At least one option must be selected`.
If the input format is wrong or SMILES cannot be compiled correctly, the result pages will display appropriate error messages. 

The results are displayed in a pandas html table format, and this result can be exported as an excel file by clicking `Click to Export to Excel` button in the result page.

Test this by entering the following input:
```
C1=CC=CC=C1_CCC1=CC=CC=C1_500
C1=CC=CC=C1_CCC1=CC=CC=C1_1000
O_[H][H]_300
O_[H][H]_400
O_[H][H]_640
C1CCCCC1_CC1CCCCC1_300
C1CCCCC1_CC1CCCCC1_400
C1CCCCC1_CC1CCCCC1_530
C[Si]1(O[Si](O[Si](O[Si](O1)(C)C)(C)C)(C)C)C_CCC(F)(Cl)I_450
CC(C)=O_c1ccc2c(c1)[nH]c1c2cccc1_300
CC(C)=O_c1ccc2c(c1)[nH]c1c2cccc1_500
CCCCCCCCCCCCCCCCCCCCCCCC_CC1CCCCC1_530
ccc
cc_ll_ll
el_ll
```
and choose any options you want. You will get a table showing results below.

![image](https://user-images.githubusercontent.com/17207530/126549532-bcc5db19-1551-4ce1-bfd3-6d9172876478.png)

jonwzheng commented on rmgweb/database/views.py:
```diff
@@ -590,6 +1039,284 @@ def parseSoluteDataComment(comment):
+    for smiles in solute_smiles_list:
+        # initialize the parameter values
+        Error = '-'
+        Comment = '-'
```
Same as above for "Error", shall we use a different & lowercase name for the comment? Like solvation_comment?

yunsiechung replied:
They are replaced with `error_msg` and `solute_comment` in this commit 3cb7caf.

jonwzheng commented on rmgweb/database/views.py:
```diff
@@ -590,6 +1039,284 @@ def parseSoluteDataComment(comment):
+        elif solute_estimator == 'SoluteML':
+            solute_epi_unc_dict = {}
+            try:
+                avg_pre, epi_unc, valid_indices = SoluteML_estimator([[smiles]])
```
This line always fails. Thus, SoluteML as the solute parameter estimation method always fails to parse the SMILES; for instance, try CC. 
`NameError: name 'rdNormalizedDescriptors' is not defined`
Probably you have to import something else from RDKit.

jonwzheng replied:
One note (for anyone reading this PR) - this error was obscured because the "except" block does not specify what exception to catch.
Thus, when we had a NameError, this too was treated normally as if there were no problem.
It is generally good practice to specify the exception we expect to catch. The problem I guess in this case is that it's not really obvious what exceptions we should expect.

yunsiechung replied:
ah I know why it fails. It's because we need to install one more package that cannot be installed through `environment.yml`.
Sorry, I forgot to add that.

You have to install one more package using the following line:
`pip install git+https://github.com/bp-kelley/descriptastorus`
But this is not available through conda install, so it cannot be added to `environment.yml`. Any idea how to add this information to this PR? Other git repo like `chemprop` just mentions to run that line when installing `chemprop`.

yunsiechung replied:
Regarding your second comment:
I actually have no idea what kind of exception to expect from `SoluteML_estimator`. I have never encountered any error except that name error. I can add the case for `NameError` exception. But besides that, we may never encounter any other errors, so I will just make the rest as general exception.

jonwzheng replied:
(1) Perhaps we can add it to the .README and to [the wiki](https://github.com/ReactionMechanismGenerator/RMG-Py/wiki/Setting-up-RMG-website). Other than that I am not sure what would be best. I would imagine this is an easy step to miss
(2) I agree that it's not clear what exceptions to expect during "normal operation," and it's probably fine that we just leave it as 'except:'. I did see that if we remove the `except` block we encounter an  `ArgumentError` if the SMILES string cannot be parsed. And then there might be `IndexError` indexing into  `epi_unc`.

yunsiechung replied:
I was getting `ValueError` instead of `NameError` you got when I uninstalled the `descriptastorus` package to reproduce the error you got. Therefore, I instead catch this type of error by looking for some keywords in the error message as shown in this commit 4b8ca68 and display the following error page with installation instruction:

![descriptastorus_error_msg](https://user-images.githubusercontent.com/17207530/131521784-0665e676-cb9a-400d-821a-ee2639df0c82.png)
 
I also added the installation guide to the .README and to the wiki.

jonwzheng commented on rmgweb/database/views.py:
```diff
@@ -590,6 +1039,284 @@ def parseSoluteDataComment(comment):
+            solute_epi_unc_dict = {}
+            try:
+                avg_pre, epi_unc, valid_indices = SoluteML_estimator([[smiles]])
+                solute_data.E = avg_pre[0][0]
```
You might be able to rewrite this block pythonically. I'm not 100% sure on the implementation but I'm imagining something like
```
[solute_data.E, solute_data.S, solute_data.A, solute_data.B, solute_data.L] = (avg_pre[0][i] for i in range(5)) 
```
Same for below, but double check that this works though.

yunsiechung replied:
Thanks. I did it in this commit f82f20f in a new function `get_solute_data_from_SoluteML`.

jonwzheng commented on rmgweb/database/views.py:
```diff
@@ -590,6 +1039,284 @@ def parseSoluteDataComment(comment):
+        # check whether we have all solvent parameters to calculate dGsolv and dHsolv
+        abraham_parameter_list = [solvent_data.s_g, solvent_data.b_g, solvent_data.e_g, solvent_data.l_g,
+                                  solvent_data.a_g, solvent_data.c_g]
+        dGsolv_availble = not any(param is None for param in abraham_parameter_list)
```
typo? should be "available" for this and the `dHsolv_available` below.

yunsiechung replied:
Thanks for catching this. It's fixed in 3cb7caf.

jonwzheng commented on rmgweb/database/views.py:
```diff
@@ -590,6 +1039,284 @@ def parseSoluteDataComment(comment):
+                    dHsolv298 = 'not available'
+                if dGsolv_availble and dHsolv_availble:
+                    dSsolv298 = (dHsolv298 - dGsolv298) / 298  # has the unit of dGsolv298 divided by Kelvin
+                    if abs(dSsolv298) < 1000:
```
Maybe you can write a function that generally formats a number using these repeated criteria:
- if <1000 then display normally, otherwise use sci notation
- maybe divide by specified T (default = 298)
- maybe do unit conversions

Then you can just call that function each time instead of repeating them here.

yunsiechung replied:
I created a new function `clean_up_value` for this now.

jonwzheng commented on rmgweb/database/views.py:
```diff
@@ -2607,43 +3334,211 @@ def moleculeSearch(request):
+        form = SolvationSearchMLForm(initial, error_class=DivErrorList)
+        if posted.is_valid():
+            solvent_solute_smiles = posted.cleaned_data['solvent_solute_smiles']
+            calc_dGsolv = posted.cleaned_data['calc_dGsolv']
```
Instead of unpacking the dict, you could just directly pass the dictionary values into `get_solvation_data_ML`. As pseudocode, something like:
```
if posted.isvalid():
     calc_dict = posted.cleaned_data
     if 'submit' in request.POST:
          solvation_data_results, additional_info_list = get_solvation_data_ML(calc_dict['solvent_solute_smiles'], calc_dict['calc_dGsolv'] ... etc.
```

yunsiechung replied:
This is no longer needed because we decided to pass the input arguments instead of the output arguments.

jonwzheng reviewed (commented):
Awesome work! This is a huge step forward for the solvation capabilities on the website (thousands of new lines of code!!), and the implementation looks very good. Here are my overall thoughts so far.... BTW, I updated the dev website w/ this branch so you can test it there as well.

(1) Solvent Search
- If there is only one result, maybe the search page can just directly bring the user to the solvent entry in the solvent database.

(2) Solute & Solvation LSER Search
- Have you considered a feature to compare multiple calculation methods? 

(3) Overall
- The URL for searches is really long. Though the explicit limit is 100 molecules, in practice one can only do a search involving ~30 molecules - after this, the URL becomes too long for the server to handle. Even below this limit, the URLs are so long that they are hard to work with. Perhaps you can just send the inputs instead as the http request, instead of the full results - I see that you intentionally avoided this to avoid double calculation for exporting to Excel but  maybe there is a way to avoid that? We can look at it together. 
- In the past, the website has had issues when people repeatedly click on buttons to do lengthy calculations, which takes up server resources & can crash the site in extreme cases. To mitigate this, we have a loading message  for kinetics database searches. For the bulk ML searches, to be safe, it would be good to put a loading message like in PR #207 so people don't accidentally overload the server. 
- Could we add a header to the Solvation Tools page (for stylistic consistency)?
- Very minor point: The tables could be formatted a bit more aesthetically. Maybe a line break after "Click to export to excel"; Centering of column headers + perhaps some padding, so it looks better.
- I see that the convention you have used is to include the entry number every time a solvent is displayed (e.g. "Result # 1 - 1. water" or "Result # 2 - 27. methanol_50_water_50"). I think there is a lot of merit in doing it this way; e.g. for solvent mixtures or esoteric solvents, if you start with a Solvent Search, then you will know exactly how far a dropdown menu to search for your solvent; otherwise it's not obvious where to find methanol_50_water_50 in the list. Yet I have also thought about removing the entry # for two reasons - (a) aesthetic consistency & avoiding confusion with extraneous numbers; (b) When I open a dropdown box, often I type the solvent name to highlight it (E.g. if I know I want to find the solvent "methanol", I will type "meth" to bring me to methanol without finding it in the list). If everything starts with a number, I can no longer reliably do this.  But I also do like that the solvent numbering helps the user find more complicated solvents in the dropdown menu. What are your thoughts?  [ Perhaps in the dropdown menus for solvent selection, the entry # can be appended, rather than prepended. ]
- There are large chunks of code that have been commented out in `views.py`, feel free to delete them since they're pretty big and superseded by new functions.

yunsiechung commented on rmgweb/database/templates/solvationSolventData.html:
```diff
@@ -0,0 +1,52 @@
+<h3>The given solvent is not found in the RMG solvent library.</h3>
+{% else %}
+{% for solventLabel, solventEntry, href in solventInfoList %}
+<h3>Result #{{ forloop.counter }} &ndash; <a href="{{ href }}">{{ solventEntry.index }}. {{ solventLabel }}</a></h3>
```
okay! I will do that

yunsiechung commented on rmgweb/database/views.py:
```diff
@@ -427,114 +446,544 @@ def solvationEntry(request, section, subsection, index):
+        # initialization
+        solvent_smiles = '-'
+        solute_smiles = '-'
+        Error = '-'
```
I will replace it with `error_msg`

yunsiechung replied:
It's replaced with `error_msg` in this commit 3cb7caf.

yunsiechung commented on rmgweb/database/forms.py:
```diff
@@ -243,6 +243,172 @@ def clean_temp(self):
+    def clean_solvent_solute_smiles(self):
+        """
+        Custom validation for the solute_smiles_list field to ensure that the number of solute SMILES
+        does not exceed 100.
```
It's fixed with 5d799c9. I lowered the limitation to 50 for all solvation tools.

yunsiechung commented on rmgweb/database/views.py:
```diff
@@ -427,114 +446,544 @@ def solvationEntry(request, section, subsection, index):
+    # get predictions for each given solvent_solute SMILES
+    for solvent_solute in solvent_solute_smiles_list:
+        # initialization
+        solvent_smiles = '-'
```
The code is cleaned up with this fff2f51.

yunsiechung commented on rmgweb/database/views.py:
```diff
@@ -427,114 +446,544 @@ def solvationEntry(request, section, subsection, index):
+            dSsolv298 = "{:.5f}".format(dSsolv298)  # round to 5 decimal places for solvation entropy
+
+        # append the results
+        solvation_data_results['Input'].append(solvent_solute)
```
I changed the way the results are appended in these commits:
f82f20f, fff2f51, and 4a96f9c.

yunsiechung commented on rmgweb/database/views.py:
```diff
@@ -427,114 +446,544 @@ def solvationEntry(request, section, subsection, index):
+            if calc_dSsolv and dGsolv298 != '-' and dHsolv298 != '-':
+                dSsolv298 = (dHsolv298 - dGsolv298) / 298  # default is in kcal/mol/K
+            if calc_logK and dGsolv298 != '-':
+                logK = -dGsolv298 * 4184 / (
```
I chose not to use the user's unit preference setting because I think many users will prefer using these tools without having to create an account. Maybe I can also use the user's unit preference setting in the future, but I won't do it for this PR.
Instead, I created a function `convert_energy_unit` that does the unit conversion for the energy in this commit f82f20f.

yunsiechung commented on rmgweb/database/views.py:
```diff
@@ -427,114 +446,544 @@ def solvationEntry(request, section, subsection, index):
+
+    dGsolv_required = any(
+        [calc_dGsolv, calc_dSsolv, calc_logK, calc_logP])  # whether or not dGsolv calculation is needed
+    dHsolv_requied = any([calc_dHsolv, calc_dSsolv])  # whether or not dHsolv calculation is needed
```
Thanks for catching this. It's fixed in 3cb7caf.

yunsiechung commented on rmgweb/database/views.py:
```diff
@@ -427,114 +446,544 @@ def solvationEntry(request, section, subsection, index):
+            solvent_smiles = pair_list[0]
+            solute_smiles = pair_list[1]
+            pair_smiles = [[solvent_smiles, solute_smiles]]
+            if dGsolv_required:
```
They use different functions for calculations and I use different variables for each calculation results, so iterating through them was not an ideal option for me. But I made the function look cleaner and shorter overall by dividing them into separate, different functions. e.g. Now this is done using `get_solvation_from_DirectML`.

yunsiechung commented on rmgweb/database/views.py:
```diff
@@ -427,114 +446,544 @@ def solvationEntry(request, section, subsection, index):
+                    if Error == '-':
+                        Error = 'Temperature is out of range'
+                    else:
+                        Error += ', Temperature is out of range'
```
As you recommended, I broke this up into multiple functions 4a96f9c.

yunsiechung commented:
@jonwzheng Sorry for taking it so long, but I finally made nearly all the changes you suggested.
I replied to your previous individual comments. 

Regarding your general comments:
(1) Solvent Search 
-  I changed the formats a bit for displaying the solvent search result. Instead of displaying all the solvent parameters on the result page, it now only gives the list of matched solvents and the link to their solvent database entry like this:

![image](https://user-images.githubusercontent.com/17207530/131523203-cc2747aa-ce3b-441b-a87f-0099c78f1bda.png)

(2) Solute & Solvation LSER Search
- I plan to add a feature that can compare multiple calculation methods, but I won't do it in this PR because it's already too long. I will add it in the future!

(3) Overall

- **Regarding the URL for searches being too long**: now, the URL only contains the input arguments and I limited the input to 50 solutes, solvent-solute, or solvent-solute-temperatures, so the URL won't be too long any longer. As we discussed in private, in order to avoid double calculation for exporting to Excel, the results from html is passed as a hidden variable when the excel export option is selected.
- **Regarding displaying a loading image and message**: this is now implemented in this commit a666fd3.
- **Regarding the missing header for the Solvation Tools page**: it is added in f90cb13.
- **Regarding formatting the result page**: I'm satisfied with the current formats of the result page and table, so I didn't make any changes. As you said, this is minor, so the formatting can be done in the future if needed.
- **Regarding showing the entry number for a solvent**: As we discussed in private, I think it's easier to look up a solvent by entry number and the user is supposed to look up a solvent using the `Solvent Search` tool anyway, which should display the entry number. So I didn't change the format for this.
- **Regarding the blocks of commented codes that are no longer used**: these are now deleted 0b2932c.


My overall comment:
I made lots of new commits, but they are just mostly moving around the code to make it look cleaner and nicer. None of the results should have changed, and I haven't experienced any error when I tested with many different inputs. It would be nice if this PR can be reviewed and merged soon!

yunsiechung commented:
@jonwzheng I squashed the fixup commits and force-pushed. This is ready to be merged whenever you want!
